channeldb/payments.go

```diff
@@ -252,9 +252,70 @@ type Payment struct {
 	Failure *FailureReason
 }
 
+// ToMPPayment converts a legacy payment into an MPPayment.
+func (p *Payment) ToMPPayment() *MPPayment {
+	var (
+		htlcs   []HTLCAttempt
+		settle  *HTLCSettleInfo
+		failure *HTLCFailInfo
+		reason  *FailureReason
+	)
+
+	// Promote the payment preimage to proper settle struct, if it exists.
+	if p.Preimage != nil {
+		// NOTE: SettleTime is not set for legacy payments.
+		settle = &HTLCSettleInfo{
+			Preimage: *p.Preimage,
+		}
+	}
+
+	// Promote the payment failure to a proper fail struct, if it exists.
+	if p.Failure != nil {
+		// NOTE: FailTime is not set for legacy payments.
+		failure = &HTLCFailInfo{}
+		reason = p.Failure
+	}
+
+	// Either a settle or a failure may be set, but not both.
+	if settle != nil && failure != nil {
+		panic("htlc attempt has both settle and failure info")
+	}
+
+	// Populate a single HTLC on the MPPayment if an attempt exists on the
+	// legacy payment. If none exists we will leave the attempt info blank
+	// since we cannot recover it.
+	if p.Attempt != nil {
+		// NOTE: AttemptTime is not set for legacy payments.
```

Comment:
The checks for `p.Preimage`  and `p.Failure`, can they be moved inside this if?

Reply:
turns out the preimage can, but failure needs to be outside. we have have a failure reason w/o an attempt

Reply:
I think timeout can have happened even without an attempt. That being said, we must double check that legacy-legacy payment was given an attempt during migration. In any case I don't think we should assume it will always be present even if we have success/failure

Reply:
@halseth not sure I follow, does that change anything about my last comment?

Reply:
No, I was just attempting to agree with you 🤣  I guess maybe I was confused by "have have"

Reply:
ah okay sorry, fixed the have have 😂